jetbrains: verify download urls

herberteuler opened the pull request:
## Description of changes

A new version of MPS, `2023.3.1`, contains three numbers in its version. The update script currently cannot handle this case and generates an invalid URL:

```
(Generated by the current script)
https://download.jetbrains.com/mps/2023.3.1/MPS-2023.3.1.tar.gz

(The real URL)
https://download.jetbrains.com/mps/2023.3/MPS-2023.3.1.tar.gz
```

This PR improves the update script to verify the URLs to support patch versions.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

herberteuler committed:
jetbrains: 2023.3.0 -> 2024.1.3

jetbrains.dataspell: 2024.1.2 -> 2024.1.3
jetbrains.mps: 2023.3.0 -> 2023.3.1
jetbrains.rust-rover: 2024.1.1 -> 2024.1.2

herberteuler committed:
jetbrains.plugins: update

herberteuler committed:
jetbrains: verify download urls

zlepper commented:
Just out of curiosity: has anyone reached out to jetbrains to check if this change was intentional or an accident? Because if this is intentional we might want to prepare for it to happen for other products too, if not they might fix the urls on their side?

herberteuler commented:
Two things are worth mentioning: a) They have changed the URLs for all MPS versions, not just for the most recent version. b) The [schemes](https://www.jetbrains.com/idea/download/other.html) [for](https://www.jetbrains.com/pycharm/download/other.html) [other](https://www.jetbrains.com/webstorm/download/other.html) [products](https://www.jetbrains.com/clion/download/other.html) are different from that for MPS. In fact, `versionMajorMinor` is solely used for MPS in the templates right now.

Besides, this change to the script is generic enough to support both the old and new schemes and does not restrict it to MPS.

herberteuler commented:
@zlepper I had a misunderstanding. It was not because they changed the scheme but rather because this is a new version of MPS with three numbers in it (2023.3.1). It is amazing how this has never happened in the past to trigger an error in the update script.

herberteuler commented:
I have updated the description.

zlepper commented:
Result of `nixpkgs-review pr 320186` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>jetbrains.dataspell</li>
    <li>jetbrains.rust-rover</li>
  </ul>
</details>

zlepper commented on pkgs/applications/editors/jetbrains/bin/update_bin.py:
```diff
@@ -57,6 +59,19 @@ def download_sha256(url):
 channels = download_channels()
 
 
+def get_url(template, version_or_build_number, version_number):
+    parts = version_number.split(".")
+    for k in range(len(parts), 0, -1):
+        s = ".".join(parts[0:k])
+        url = template.format(version=version_or_build_number, versionMajorMinor=s)
+        try:
+            if urllib.request.urlopen(url).getcode() == 200:
+                return url
+        except HTTPError:
+            pass
+    return None
```
I'm barely a python noob, so take all my feedback on the python with a very large grain on salt. 

Should we consider usign the Version type from `from packaging import version` (Line 12) to parse the versions rather than dealing with string manipulation ourselves? Especially for EAP packages which has slightly strange version numbers?

herberteuler replied:
Yes, that makes sense. I am a casual Python user, too, and I was unaware of the `version` module. Thanks.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1757

herberteuler commented:
@leona-ya Could you please also take a look at this PR. It fixes the URL issue in #319290.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1832

herberteuler commented:
Please notify me when someone is ready to merge this. I will resolve the conflicts only then in case of there are further needs to resolve future conflicts.

leona-ya commented:
I would merge this, can you rebase please? Sorry for the delay

herberteuler committed:
Merge remote-tracking branch 'upstream/master' into fix/jetbrains-download-url

herberteuler committed:
jetbrains: 2024.1 EAP -> 2024.1.6

jetbrains.pycharm-community: 2024.1.3 -> 2024.1.4
jetbrains.pycharm-professional: 2024.1.3 -> 2024.1.4
jetbrains.rust-rover: 2024.1.2 -> 2024.1.6
jetbrains.writerside: 2024.1 EAP -> 2024.1 EAP

herberteuler committed:
jetbrains.plugins: update

herberteuler commented:
> I would merge this, can you rebase please? Sorry for the delay

Done. Please take a look. Thanks.

leona-ya commented:
Can you remove the merge commit, and rebase your branch on `master ` instead? We don't use merge commits within PRs here

herberteuler commented:
OK. I will create a new PR to include only the changes to `update_bin.py`.
